### PR TITLE
git-smash: init at 0.1.1

### PR DESCRIPTION
**Pull request by bcyran**
Adding [`git-smash`](https://github.com/anthraxx/git-smash) - a tool for quick amending of previous commits. Automates the stage, fixup commit, interactive rebase workflow.

FYI @anthraxx

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/gi/git-smash/package.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  nativeBuildInputs = [
+    installShellFiles
+    makeWrapper
+  ];
+
+  postFixup = ''
+    wrapProgram "$out/bin/git-smash" --prefix PATH : "${fuzzyMatcherPackage}/bin"
+  '';
+
+  postInstall = ''
```
Comment (Aleksanaa) [inline pkgs/by-name/gi/git-smash/package.nix:34]:
```suggestion
  postInstall = lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
```

File: pkgs/by-name/gi/git-smash/package.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
+{
+  lib,
+  fetchFromGitHub,
+  rustPlatform,
+  installShellFiles,
+  makeWrapper,
+  fuzzyMatcherPackage ? fzf,
```
[the remark this answers is not in the document]
Comment (Aleksanaa) [reply pkgs/by-name/gi/git-smash/package.nix:7]: Not a bad idea, better to leave a comment here

[the remark this answers is not in the document]
Comment (bcyran) [reply pkgs/by-name/gi/git-smash/package.nix]: Absolutely, added a comment.

File: pkgs/by-name/gi/git-smash/package.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = with lib; {
+    description = "Smash staged changes into previous commits to support your Git workflow, pull request and feature branch maintenance";
+    homepage = "https://github.com/anthraxx/git-smash";
+    changelog = "https://github.com/anthraxx/git-smash/blob/${src.rev}/CHANGELOG.md";
+    license = licenses.mit;
+    mainProgram = "git-smash";
+    maintainers = with maintainers; [ bcyran ];
```
[the remark this answers is not in the document]
Comment (bcyran) [reply pkgs/by-name/gi/git-smash/package.nix]: Corrected.

File: pkgs/by-name/gi/git-smash/package.nix (hunk 1 above)
Comment (bcyran) [reply pkgs/by-name/gi/git-smash/package.nix]: Changed as proposed. Just to make sure I understand what's going on. This is to make sure we won't try to run the `git-smash` binary on an unsupported platform when cross compiling. Is that correct?

File: pkgs/by-name/gi/git-smash/package.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  cargoHash = "sha256-omITZMBWzYlHHim/IXNa1rtiwHqpgLJ5G9z15YvDRi0=";
+
+  nativeBuildInputs = [
+    installShellFiles
+    makeWrapper
+  ];
+
+  postFixup = ''
+    wrapProgram "$out/bin/git-smash" --prefix PATH : "${lib.makeBinPath [ fuzzyMatcherPackage ]}"
```
Comment (SuperSandro2000) [inline pkgs/by-name/gi/git-smash/package.nix:32]:
```suggestion
    wrapProgram "$out/bin/git-smash" --suffix PATH : "${lib.makeBinPath [ fzf ]}"
```

File: pkgs/by-name/gi/git-smash/package.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
+{
+  lib,
+  fetchFromGitHub,
+  rustPlatform,
+  installShellFiles,
+  makeWrapper,
+  fuzzyMatcherPackage ? fzf, # can use either fzf or skim as a fuzzy matcher
```
Comment (SuperSandro2000) [inline pkgs/by-name/gi/git-smash/package.nix:7]:
```suggestion
```
we should just rely on PATH and not trigger useless rebuilds like this

Comment (bcyran) [reply pkgs/by-name/gi/git-smash/package.nix]: Ok then, changed as proposed. I don't really get the "trigger useless rebuilds" part. Could you please explain this?

**Review by Aleksanaa — APPROVED** [review]

**Comment by anthraxx** [comment]
Thanks everyone for working on this. It's much appreciated and motivating.

**Pull request merged**